tempesta_fw/ss_skb.c

```diff
@@ -816,6 +816,50 @@ __skb_fragment(struct sk_buff *skb_head, struct sk_buff *skb, char *pspt,
 			ret = __split_pgfrag(skb_head, skb, i, offset, len, it, fragn);
 			goto done;
 		}
+		/*
+		 * In case another fragment does not exist, remember the
+		 * fragment number, after which you need to insert/delete data.
+		 */
+		if (offset == d_size)
+			j = i;
+	}
+
+	if (j >= 0) {
+		skb_frag_t *frag = &si->frags[j + 1];
+
+		if (len > 0) {
+			ret = __new_pgfrag(skb_head, skb, len, j + 1, 1);
+			__it_next_data(skb, j + 1, it, fragn);
+			goto done;
+		}
+
```

Comment:
When we can have the negative `len` and `offset == d_size` above? The case looks wrong for me.

I tried to run your previous code version with the small patch
```
@@ -821,7 +821,7 @@ __skb_fragment(struct sk_buff *skb_head, struct sk_buff *skb, char *pspt,
 		 * fragment number, after which you need to insert a new one.
 		 */
 		if (offset == d_size) {
-			len = max_t(long, len, offset - d_size);
+			BUG_ON(len < 0);
 			j = i;
 		}
 	}
```
It doesn't hit the assertion, but it seems there is some memory leak - the full functional tests suite fails with OOM while it passes on the master as of today.